File: app/lib/features/scanner/domain/document_scan.dart
/// Status of a document scan.
enum ScanStatus {
  uploading,
  processing,
  completed,
  failed;

  static ScanStatus fromString(String value) {
    switch (value) {
      case 'uploading':
        return ScanStatus.uploading;
      case 'processing':
        return ScanStatus.processing;
      case 'completed':
        return ScanStatus.completed;
      case 'failed':
        return ScanStatus.failed;
      default:
        return ScanStatus.processing;
    }
  }
}

/// Represents a document scan result.
class DocumentScan {
  const DocumentScan({
    required this.id,
    required this.status,
    this.fileUrl,
    this.ocrText,
    this.translation,
    this.explanation,
    this.documentType,
    this.createdAt,
  });

  final String id;
  final ScanStatus status;
  final String? fileUrl;
  final String? ocrText;
  final String? translation;
  final String? explanation;
  final String? documentType;
  final DateTime? createdAt;

  factory DocumentScan.fromJson(Map<String, dynamic> json) {
    return DocumentScan(
      id: json['id'] as String,
      status:
          ScanStatus.fromString(json['status'] as String? ?? 'processing'),
      fileUrl: json['file_url'] as String?,
      ocrText: json['ocr_text'] as String?,
      translation: json['translation'] as String?,
      explanation: json['explanation'] as String?,
      documentType: json['document_type'] as String?,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at'] as String)
          : null,
    );
  }
}

File: backend/ai_service/models/document_scan.py
"""DocumentScan model — maps to the document_scans table (DATA_MODEL.md §9)."""

import uuid
from datetime import datetime

from sqlalchemy import DateTime, ForeignKey, Integer, String, Text, func
from sqlalchemy.orm import Mapped, mapped_column

from database import Base


class DocumentScan(Base):
    """Document scan record — user-uploaded document OCR + translation."""

    __tablename__ = "document_scans"

    id: Mapped[str] = mapped_column(
        String(36), primary_key=True, default=lambda: str(uuid.uuid4())
    )
    user_id: Mapped[str] = mapped_column(
        String(128), ForeignKey("profiles.id"), nullable=False
    )
    file_url: Mapped[str] = mapped_column(Text, nullable=False)
    file_name: Mapped[str] = mapped_column(String(255), nullable=False)
    file_size_bytes: Mapped[int] = mapped_column(Integer, nullable=False)
    ocr_text: Mapped[str | None] = mapped_column(Text, nullable=True, default=None)
    translation: Mapped[str | None] = mapped_column(
        Text, nullable=True, default=None
    )
    explanation: Mapped[str | None] = mapped_column(
        Text, nullable=True, default=None
    )
    document_type: Mapped[str | None] = mapped_column(
        String(50), nullable=True, default=None
    )
    source_language: Mapped[str] = mapped_column(
        String(5), nullable=False, default="ja"
    )
    target_language: Mapped[str] = mapped_column(String(5), nullable=False)
    status: Mapped[str] = mapped_column(
        String(20), nullable=False, default="uploading"
    )
    error_message: Mapped[str | None] = mapped_column(
        Text, nullable=True, default=None
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now()
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, server_default=func.now(),
        onupdate=func.now(),
    )
    deleted_at: Mapped[datetime | None] = mapped_column(
        DateTime(timezone=True), nullable=True, default=None
    )
